src/storage.rs

```diff
@@ -0,0 +1,533 @@
+            .find(|c| c.id == data_proposal.pos && c.txs == data_proposal.inner);
+
+        match car {
+            None => {
+                warn!(
+                    "Vote for Car that does not exist! ({sender}) data_proposal: {} / lane: {}",
+                    data_proposal, self.lane
+                );
+                None
+            }
+            Some(v) => {
+                v.votes.insert(sender.clone());
+                Some(v.votes.clone())
+            }
+        }
+    }
+
+    fn update_other_votes(
+        &mut self,
+        sender: &ValidatorPublicKey,
+        data_proposal: &DataProposal,
+        voters: HashSet<ValidatorPublicKey>,
+    ) -> Option<Vec<ValidatorPublicKey>> {
+        if let Some(lane) = self.other_lanes.get_mut(sender) {
+            let car = lane
+                .cars
+                .iter_mut()
+                .find(|c| c.id == data_proposal.pos && c.txs == data_proposal.inner);
+
+            match car {
+                None => {
+                    warn!(
+                        "Update PoA for Car that does not exist! ({sender}) data_proposal: {} / lane: {}",
+                        data_proposal, self.lane
+                    );
+                    return None;
+                }
+                Some(v) => {
+                    v.votes.extend(voters);
+                    return Some(v.votes.clone().into_iter().collect());
+                }
+            }
+        }
+        None
+    }
+
+    fn append_data_proposal(
+        &mut self,
+        sender: &ValidatorPublicKey,
+        data_proposal: &DataProposal,
+    ) -> bool {
+        let lane = self.other_lanes.entry(sender.clone()).or_default();
+        let tip = lane.current_mut();
+
+        let mut tip_id: Option<usize> = None;
+
+        // merge parent votes
+        if let Some(p) = tip {
+            tip_id = Some(p.id);
+            if let Some(v) = data_proposal.parent_poa.as_ref() {
+                p.votes.reserve(v.len());
+                for v in v.iter() {
+                    p.votes.insert(v.clone());
+                }
+            }
+        }
+
+        if data_proposal.parent == tip_id {
+            lane.cars.push(Car {
+                id: tip_id.unwrap_or(0) + 1,
+                parent: tip_id,
+                txs: data_proposal.inner.clone(),
+                votes: HashSet::from([self.id.clone(), sender.clone()]),
+            });
+            true
+        } else {
+            false
+        }
+    }
+
+    fn has_data_proposal(
+        &mut self,
+        sender: &ValidatorPublicKey,
+        data_proposal: &DataProposal,
+    ) -> bool {
+        let found = self
+            .other_lanes
+            .entry(sender.clone())
+            .or_default()
+            .cars
+            .iter()
+            .any(|c| c.id == data_proposal.pos && c.txs == data_proposal.inner);
+
+        if !found {
+            warn!(
+                "Data proposal {} not found in lane {}",
```

Comment:
I ran 2 nodes and got a lot of warnings here, is it intended if there is no data ? If it is "normal" to have `!found` this log should be a debug and not a warn please